pgjdbc/src/main/java/org/postgresql/jdbc/PgClob.java

```diff
@@ -5,62 +5,153 @@
 
 package org.postgresql.jdbc;
 
+import org.postgresql.util.GT;
+import org.postgresql.util.LimitedReader;
+import org.postgresql.util.PSQLException;
+import org.postgresql.util.PSQLState;
+
+import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
 import java.io.Reader;
+import java.io.StringWriter;
+import java.io.Writer;
 import java.nio.charset.Charset;
 import java.sql.Clob;
+import java.sql.NClob;
 import java.sql.SQLException;
 
-public class PgClob extends AbstractBlobClob implements java.sql.Clob {
+class PgClob extends AbstractBlobClob implements Clob, NClob {
+
+  PgClob(org.postgresql.core.BaseConnection conn) throws java.sql.SQLException {
+    super(conn);
+  }
 
-  public PgClob(org.postgresql.core.BaseConnection conn, long oid) throws java.sql.SQLException {
+  PgClob(org.postgresql.core.BaseConnection conn, long oid) throws java.sql.SQLException {
     super(conn, oid);
   }
 
-  public synchronized Reader getCharacterStream(long pos, long length) throws SQLException {
-    checkFreed();
-    throw org.postgresql.Driver.notImplemented(this.getClass(), "getCharacterStream(long, long)");
+  private int getBufferSize(long length) {
+    return (int) Math.min(length * 1.2, 64000);
   }
 
+  @Override
   public synchronized int setString(long pos, String str) throws SQLException {
-    checkFreed();
-    throw org.postgresql.Driver.notImplemented(this.getClass(), "setString(long,str)");
+    // TODO: pos>0 check
+    return setString(pos, str, 0, str.length());
+  }
+
+  private long measureBytes(long numCharacters) throws SQLException {
+    if (numCharacters == 0) {
+      return 0;
+    }
+    // TODO: offload to backend somehow?
+    Reader characterStream = getCharacterStream(1, numCharacters);
+    int bufferSize = (int) (numCharacters < 8192 ? numCharacters * 1.2 : numCharacters);
+    char[] tmp = new char[bufferSize];
+    int read;
+    long len = 0;
+    try {
+      while ((read = characterStream.read(tmp)) != -1) {
+        len = len + read;
+      }
+    } catch (IOException e) {
+      throw new PSQLException(GT.tr("Unable to skip {0} characters, large object oid is {1}",
+          numCharacters, getOid()),
+          PSQLState.UNEXPECTED_ERROR, e);
+    }
+    return len;
```

Comment:
This is numbers of chars, not bytes, right?

Reply:
For `clob` most of the numbers are chars.
This is chars.

I'm not sure how emojis should count though, so I use the same meaning as `String#charAt`

Reply:
But method name is measureBytes, which seems a bit misleading if actually measuring chars.

In looking at jdbc javadoc, I read it to count chars, not codepoints.

Reply:
`private long measureBytes(long numCharacters) throws SQLException {`

It measures the number of bytes that consume first numCharacters chars of the lob data.

Reply:
Perhaps I am just being dense. Wouldn't measuring bytes require a specific character encoding? If the encoding is ascii or iso-8859-1 then this would be accurate. But if the encoding is utf-8, then any of the fun latin diacritics could lead to 1 char being 2 bytes. Moving to other languages makes that even more likely.

Reply:
I've absolutely no idea why `java.sql.Clob` is mapped to `Large Object` API, however the implementation there always was like "large object + client encoding"

We might want to map `java.sql.Clob` to `text` datatype, however it would definitely break existing apps.

Reply:
Max value of `text` is 1Gb

Reply:
I seem to have lost the plot. How did we get on the text vs LO subject ?

Reply:
I think the mapping to LO is good.
I just think that counting bytes (presumably of data stored in LO) and counting chars (in java representation of clob) are 2 different things.

Reply:
@davecramer , the thing is as follows:

JDBC has `java.sql.Blob` for large binary content with operations like "streaming reads, streaming writes".
We can use `Large Object API` for that.

JDBC has `java.sql.Clob` for large text content as well. The same thing with "streaming reads, streaming writes" applies there.
The problem is PostgreSQL does **NOT** support `large text` data type at all.


For God knows what reasons pgjdbc uses `Large Object API` for `java.sql.Clob` as well. Of course the data is stored as binary at the backend, and pgjdbc uses `client` encoding to convert String to the binary format.

**Then goes the question**: what is the proper datatype for "large text" (==`java.sql.Clob`)?
`text` might make sense (at least the DB understands the encoding), however it has 1Gb limit.
`Large Object` is the current implementation, however it looks odd as it ties the data to the **client** encoding.

Reply:
>I wonder how many Clob's are larger than 1GB ? Of course we have no way of knowing...

Suppose we go with `text`.
Then a person needs a text that exceed 1Gb.

Then goes the funny part :)

Should we file an email thread on that?

Note: `lo`+`clob` can be made backward-compatible. For instance, if we (or PostgreSQL) prepends `lo` contents with `lo.encoding=utf-8; ...` sequence, then pgjdbc can easily tell if `lo` is "old" or "new" one.

Reply:
While it is certainly reliant on the client to apply the encoding, I thought 'connection.getEncoding()' returned the database’s encoding.
Postgresql’s lack of character LO data type certainly limits the functionality (and likely use) of a jdbc clob. As the column itself cannot support text functions, such as qualifying in a where clause.
It would probably be valuable to document how each of the methods behave.

Out of curiosity, is the text limit in bytes or characters?

Reply:
pg max column size is 1GB in bytes

Reply:
d) Enforce the presence of a valid Java system property before reading or writing from CLOB; the value of said property would be the encoding/decoding scheme to use to convert the stored binary to String.

Providing an API default of UTF-8 for that system property, instead of having it be reliant on client encoding, would be recommended.

I dislike the idea of adding additional content to the stored data that only pgJDBC can understand thus making access via libpq overly challenging.

Reply:
@polobo this seems like the path of least resistance 
+1

Reply:
"d" is the same as "a".

>I dislike the idea of adding additional content to the stored data that only pgJDBC can understand thus making access via libpq overly challenging.

Who else can understand the current `lo(clob)` contents?
Can one execute `select * from table where lob like '%abcd%'`?

The point is current format is broken. Enforcing an encoding won't help much.

Here's a couple of problems with `plain lo` for character data:

P1) Note: `Clob#length()` has to return the number of characters somehow. Currently it requires to fetch the contents to the Java side, decode the bytes and count the characters.
`String Clob#getSubString(long pos, int length)` users characters for `pos` and `length`.

On top of that, application would likely use `clob.getSubString(1, clob.length())` pattern, so it would produce multiple passes over the `lo` contents.

P2) "Append to the tail" is implemented as `java.io.Writer setCharacterStream(long pos) ` which uses characters for `pos` again. That means "append to the tail" would have to send all the bytes to the client, count the characters, then use that byte position for write.

Reply:
Any client can read the contents of a stored LO binary.

There is no way for the current server to support the text-oriented operations you list on that content.  IIUC this limitation cannot be overcome in the client.  We can either make a best-effort at implementing the API client-side or continue with "UnsupportedOperation" on the basis that indeed the backing database engine doesn't support it.

Relevant API links:

https://www.postgresql.org/docs/10/static/largeobjects.html
and
https://www.postgresql.org/docs/10/static/lo-funcs.html

I'd go for "best effort" in the client then decide whether to try and marshal resources to implement the capabilities in the server that the API can then be later modified to simply hook into.

Reply:
On client side, don’t we have access to db encoding? Could an option be to force that encoding?

Is TEXT stored in table?
When retrieving a TEXT column, is it possible to stream from server side like LO?

Reply:
>Any client can read the contents of a stored LO binary.

"read the contents" != "understand the contents".

Even `a)` / `d)` are pgjdbc-specific encodings.

> There is no way for the current server to support the text-oriented operations you list on that content.

`b` / `c` can do that. One can store metadata in the beginning of the `lo` in question and/or "metadatachunk at each 1MB", and the metadata could be used to make character operations more efficient.

Reply:
@bokken I do not believe it is possible to stream a text column. Of course TEXT is stored in a column. That said it does use TOAST for large text fields

Reply:
> If the backend does not support a clob, does it really make sense for driver to try and fake it?

Well, it could easily become "the official way of storing large text in PG".
The good thing is driver can use "pure client" implementation for existing backend versions, and it could use newer APIs later to access exactly the same data.
On top of that, newer backends could provide text-like SQL functions (e.g. `where lob_contains(lob, 'hi')` )

Reply:
> Is TEXT stored in table?

My question was ill-phrased. Large values being TOAST stores the values in a separate table and original column contains a pointer.

Perhaps the enhancement to backend is to expose TOAST to consumers more directly. Allow callers to ask for the pointer and then random access to the TOAST value.
This would allow for a very straightforward java clob implementation on TEXT. It would also provide ability for random access into `bytea` and arrays
It also allows for use in where clause.
The obvious downside is 1GB limit.

Reply:
> Well, it could easily become "the official way of storing large text in PG".

It seems a good idea to work towards that (whatever official implementation direction is) and provide as much capability in driver as possible with current db.

The fun part is probably getting consensus on what official way should be. 😁

Reply:
It will be more likely to be come the defacto standard than trying to get consensus ahead of time. That said I think we should poll hackers to see if anyone can foresee any additional information that we may want to add to the header to make it more extensible.

Reply:
One of the "easy to implement" options is to store UTF-16 coded values (Java's `char[]`). The downside would be extra space usage.